File: content/problems/0073-set-matrix-zeroes/solutions/first-row-column-markers/solution.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        bool firstRowZero = false, firstColumnZero = false;
        for (int value : matrix[0]) if (value == 0) firstRowZero = true;
        for (const auto& row : matrix) if (row[0] == 0) firstColumnZero = true;
        for (int row = 1; row < (int)matrix.size(); ++row) for (int column = 1; column < (int)matrix[0].size(); ++column) if (matrix[row][column] == 0) { matrix[row][0] = 0; matrix[0][column] = 0; }
        for (int row = 1; row < (int)matrix.size(); ++row) for (int column = 1; column < (int)matrix[0].size(); ++column) if (matrix[row][0] == 0 || matrix[0][column] == 0) matrix[row][column] = 0;
        if (firstRowZero) fill(matrix[0].begin(), matrix[0].end(), 0);
        if (firstColumnZero) for (auto& row : matrix) row[0] = 0;
    }
};

File: content/problems/0073-set-matrix-zeroes/solutions/first-row-column-markers/Solution.kt
class Solution {
    fun setZeroes(matrix: Array<IntArray>) {
        var firstRowZero = matrix[0].any { it == 0 }
        var firstColumnZero = matrix.any { it[0] == 0 }
        for (row in 1 until matrix.size) for (column in 1 until matrix[0].size) if (matrix[row][column] == 0) { matrix[row][0] = 0; matrix[0][column] = 0 }
        for (row in 1 until matrix.size) for (column in 1 until matrix[0].size) if (matrix[row][0] == 0 || matrix[0][column] == 0) matrix[row][column] = 0
        if (firstRowZero) matrix[0].fill(0)
        if (firstColumnZero) for (row in matrix) row[0] = 0
    }
}
